File: app/src/main/java/com/rowland/loading/views/LoadingView.java
package com.rowland.loading.views;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.PointF;
import android.graphics.Rect;
import android.graphics.RectF;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.view.View;

import com.rowland.customviews.R;
import com.rowland.loading.animations.LoadingArcAnimation;
import com.rowland.loading.utils.ArcUtils;

/**
 * Created by Rowland on 10/21/2017.
 */

public class LoadingView extends View {

    private Bitmap iconBitmap;

    // Values/Dimensions
    private float iconScaleFactor;
    private float circleRadius;
    private PointF circleCentrePoint;
    private RectF iconBitmapBounds;

    private Paint arcPaint;
    private Paint bitmapPaint;
    private Paint bitmapBackgroundPaint;
    private Paint circlePaint;

    private float mStartAngleArcOne;
    private float mStartAngleArcTwo;
    private float mSweepAngle;

    public LoadingView(Context context) {
        super(context);
        init();
    }

    public LoadingView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    private void init() {
        // Other resources
        iconBitmap = BitmapFactory.decodeResource(getResources(), R.drawable.motorbike);
        mStartAngleArcOne = 0f;
        mStartAngleArcTwo = 180f;
        mSweepAngle = 45f;
        // Dimensions
        iconScaleFactor = 0.5f;

        cfgPaint();
        cfgAnimation(100000);
    }

    private void cfgPaint() {
        // Arc Paint
        arcPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
        arcPaint.setStyle(Paint.Style.STROKE);
        arcPaint.setColor(Color.YELLOW);
        arcPaint.setStrokeWidth(25f);
        // Circle Paint
        circlePaint = new Paint(Paint.ANTI_ALIAS_FLAG);
        circlePaint.setStyle(Paint.Style.STROKE);
        circlePaint.setColor(Color.GRAY);
        circlePaint.setAlpha(50);
        circlePaint.setStrokeWidth(1f);
        //Bitmap Paint
        bitmapPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
        bitmapPaint.setStyle(Paint.Style.FILL);
        // Bitmap back
        bitmapBackgroundPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
        bitmapBackgroundPaint.setStyle(Paint.Style.FILL);
        bitmapBackgroundPaint.setColor(Color.YELLOW);
    }

    private void cfgAnimation(long duration) {
        LoadingArcAnimation animation = new LoadingArcAnimation(this, 10);
        animation.setDuration(duration);
        startAnimation(animation);
    }

    @Override
    public void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);

        int xPad = getPaddingLeft() + getPaddingRight();
        int yPad = getPaddingTop() + getPaddingBottom();

        int minW = xPad + getSuggestedMinimumWidth();
        int measuredWidth = getDefaultSize(minW, widthMeasureSpec);

        int minH = yPad + getSuggestedMinimumHeight();
        int measuredHeight = getDefaultSize(minH, heightMeasureSpec);

        setMeasuredDimension(measuredWidth, measuredHeight);
    }

    @Override
    public void onSizeChanged(int newWidth, int newHeight, int oldWidth, int oldHeight) {
        super.onSizeChanged(newWidth, newHeight, oldWidth, oldHeight);

        float xPad = getPaddingLeft() + getPaddingRight();
        float yPad = getPaddingTop() + getPaddingBottom();

        float rectifiedWidth = newWidth - xPad;
        float rectifiedHeight = newHeight - yPad;

        float xCord = rectifiedWidth / 2f;
        float yCord = rectifiedHeight / 2f;

        circleCentrePoint = new PointF(xCord, yCord);
        circleRadius = Math.min(xCord, yCord);

        float left = xCord - circleRadius / 2 * iconScaleFactor;
        float top = yCord - circleRadius / 2 * iconScaleFactor;
        float right = xCord + circleRadius / 2 * iconScaleFactor;
        float bottom = yCord + circleRadius / 2 * iconScaleFactor;
        iconBitmapBounds = new RectF(left, top, right, bottom);
    }

    @Override
    public void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        onDrawCircle(canvas);
        onDrawArcs(canvas);
        onDrawBitmap(canvas);
    }

    private void onDrawCircle(Canvas canvas) {
        canvas.drawCircle(circleCentrePoint.x, circleCentrePoint.y, circleRadius, circlePaint);
    }

    private void onDrawArcs(Canvas canvas) {
        ArcUtils.drawArc(canvas, circleCentrePoint, circleRadius, mStartAngleArcOne, mSweepAngle, arcPaint);
        ArcUtils.drawArc(canvas, circleCentrePoint, circleRadius, mStartAngleArcTwo, mSweepAngle, arcPaint);
    }

    private void onDrawBitmap(Canvas canvas) {
        double backgroundCircleRadius = Math.sqrt(Math.pow(iconBitmapBounds.height(), 2) + Math.pow(iconBitmapBounds.width(), 2)) / 2f;
        canvas.drawCircle(iconBitmapBounds.centerX(), iconBitmapBounds.centerY(), (float) (backgroundCircleRadius), bitmapBackgroundPaint);
        canvas.drawBitmap(iconBitmap, null, iconBitmapBounds, bitmapPaint);
    }

    public float[] getArcStartAngles() {
        return new float[]{mStartAngleArcOne, mStartAngleArcTwo};
    }

    public void setArcStartAngles(float[] targetStartAngles) {
        mStartAngleArcOne = targetStartAngles[0];
        mStartAngleArcTwo = targetStartAngles[1];
    }
}

File: app/src/main/java/com/rowland/loading/views/LoadingView.kt
package com.rowland.loading.views

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.PointF
import android.graphics.Rect
import android.graphics.RectF
import android.util.AttributeSet
import android.util.Log
import android.view.View

import com.rowland.customviews.R
import com.rowland.loading.animations.LoadingArcAnimation
import com.rowland.loading.utils.ArcUtils

/**
 * Created by Rowland on 10/21/2017.
 */

class LoadingView : View {

    private var iconBitmap: Bitmap? = null

    private var iconScaleFactor: Float = 0.toFloat()
    private var circleRadius: Float = 0.toFloat()
    private var circleCentrePoint: PointF? = null
    private var iconBitmapBounds: RectF? = null

    private var arcPaint: Paint? = null
    private var bitmapPaint: Paint? = null
    private var bitmapBackgroundPaint: Paint? = null
    private var circlePaint: Paint? = null

    private var mStartAngleArcOne: Float = 0.toFloat()
    private var mStartAngleArcTwo: Float = 0.toFloat()
    private var mSweepAngle: Float = 0.toFloat()

    constructor(context: Context) : super(context) {
        printLogInfo("Programmatic constructor")
        init()
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        printLogInfo("XML constructor")
        init()
    }

    private fun init() {
        // Other resources
        iconBitmap = BitmapFactory.decodeResource(resources, R.drawable.motorbike)
        mStartAngleArcOne = 0f
        mStartAngleArcTwo = 180f
        mSweepAngle = 45f
        // Dimensions
        iconScaleFactor = 0.5f

        cfgPaint()
        cfgAnimation(100000)
    }

    private fun printLogInfo(methodName: String) {
        Log.i("TAG", "$methodName called, measured size: ($measuredWidth, $measuredHeight)")
    }


    private fun cfgPaint() {
        // Arc Paint
        arcPaint = Paint(Paint.ANTI_ALIAS_FLAG)
        arcPaint!!.style = Paint.Style.STROKE
        arcPaint!!.color = Color.YELLOW
        arcPaint!!.strokeWidth = 25f
        // Circle Paint
        circlePaint = Paint(Paint.ANTI_ALIAS_FLAG)
        circlePaint!!.style = Paint.Style.STROKE
        circlePaint!!.color = Color.GRAY
        circlePaint!!.alpha = 50
        circlePaint!!.strokeWidth = 1f
        //Bitmap Paint
        bitmapPaint = Paint(Paint.ANTI_ALIAS_FLAG)
        bitmapPaint!!.style = Paint.Style.FILL
        // Bitmap back
        bitmapBackgroundPaint = Paint(Paint.ANTI_ALIAS_FLAG)
        bitmapBackgroundPaint!!.style = Paint.Style.FILL
        bitmapBackgroundPaint!!.color = Color.YELLOW
    }

    private fun cfgAnimation(duration: Long) {
        val animation = LoadingArcAnimation(this, 10)
        animation.duration = duration
        startAnimation(animation)
    }

    override fun onFinishInflate() {
        super.onFinishInflate()
        printLogInfo("onFinishInflate")
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        printLogInfo("onAttachedToWindow")
    }


    public override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        printLogInfo("onMeasure")

        val xPad = paddingLeft + paddingRight
        val yPad = paddingTop + paddingBottom

        val minW = xPad + suggestedMinimumWidth
        val measuredWidth = View.getDefaultSize(minW, widthMeasureSpec)

        val minH = yPad + suggestedMinimumHeight
        val measuredHeight = View.getDefaultSize(minH, heightMeasureSpec)

        setMeasuredDimension(measuredWidth, measuredHeight)
    }

    public override fun onSizeChanged(newWidth: Int, newHeight: Int, oldWidth: Int, oldHeight: Int) {
        super.onSizeChanged(newWidth, newHeight, oldWidth, oldHeight)
        printLogInfo("onSizeChanged")

        val xPad = (paddingLeft + paddingRight).toFloat()
        val yPad = (paddingTop + paddingBottom).toFloat()

        val rectifiedWidth = newWidth - xPad
        val rectifiedHeight = newHeight - yPad

        val xCord = rectifiedWidth / 2f
        val yCord = rectifiedHeight / 2f

        circleCentrePoint = PointF(xCord, yCord)
        circleRadius = Math.min(xCord, yCord)

        val left = xCord - circleRadius / 2 * iconScaleFactor
        val top = yCord - circleRadius / 2 * iconScaleFactor
        val right = xCord + circleRadius / 2 * iconScaleFactor
        val bottom = yCord + circleRadius / 2 * iconScaleFactor
        iconBitmapBounds = RectF(left, top, right, bottom)
    }

    override fun onLayout(changed: Boolean, left: Int, top: Int, right: Int, bottom: Int) {
        super.onLayout(changed, left, top, right, bottom)
        printLogInfo("onLayout")
    }

    public override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        printLogInfo("onDraw")
        onDrawCircle(canvas)
        onDrawArcs(canvas)
        onDrawBitmap(canvas)
    }

    private fun onDrawCircle(canvas: Canvas) {
        canvas.drawCircle(circleCentrePoint!!.x, circleCentrePoint!!.y, circleRadius, circlePaint!!)
    }

    private fun onDrawArcs(canvas: Canvas) {
        ArcUtils.drawArc(canvas, circleCentrePoint!!, circleRadius, mStartAngleArcOne, mSweepAngle, arcPaint!!)
        ArcUtils.drawArc(canvas, circleCentrePoint!!, circleRadius, mStartAngleArcTwo, mSweepAngle, arcPaint!!)
    }

    private fun onDrawBitmap(canvas: Canvas) {
        val backgroundCircleRadius = Math.sqrt(Math.pow(iconBitmapBounds!!.height().toDouble(), 2.0) + Math.pow(iconBitmapBounds!!.width().toDouble(), 2.0)) / 2f
        canvas.drawCircle(iconBitmapBounds!!.centerX(), iconBitmapBounds!!.centerY(), backgroundCircleRadius.toFloat(), bitmapBackgroundPaint!!)
        canvas.drawBitmap(iconBitmap!!, null, iconBitmapBounds!!, bitmapPaint)
    }

    var arcStartAngles: FloatArray
        get() = floatArrayOf(mStartAngleArcOne, mStartAngleArcTwo)
        set(targetStartAngles) {
            mStartAngleArcOne = targetStartAngles[0]
            mStartAngleArcTwo = targetStartAngles[1]
        }
}
